### PR TITLE
OPHJOD-1133: Add lexicalIndex to opportunity metadata

### PR DESCRIPTION
**Pull request by markus-boman-elenium**
## Description

Added sort property and Content-Language header support to /api/ehdotus/mahdollisuudet.

This change allows frontend to sort combined opportunities alphabetically by user selected language.

## Related JIRA ticket

https://jira.eduuni.fi/browse/OPHJOD-1133

### CODE REVIEW

File: src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java [hunk 1]
```diff
@@ -120,28 +135,62 @@ public List<EhdotusDto> createEhdotus(@RequestBody @Valid LuoEhdotusDto ehdotus)
...
+    return inferenceService.infer(endpoint, request, Response.class).stream()
+        .collect(Collectors.toMap(Suggestion::id, r -> r, (exising, newValue) -> exising));
+  }
+
+  private List<EhdotusDto> populateEmptyEhdotusDtos(
+      LinkedHashSet<UUID> ids, Set<UUID> tyomahdollisuusIds) {
 
+    // initialize lexicalIndex counters
+    final var tmLexicalIndex = new AtomicInteger();
+    final var kmLexicalIndex = new AtomicInteger();
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java:147]: Using AtomicInteger seems unnecessary, the stream processing is sequential so there are no thread safety issues.

File: src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java [hunk 2]
```diff
@@ -120,28 +135,62 @@ public List<EhdotusDto> createEhdotus(@RequestBody @Valid LuoEhdotusDto ehdotus)
...
-                          return new Suggestion(key, -1, tyyppi.name());
-                        }))
+            key -> {
+              var type = getTyyppi(key, tyomahdollisuusIds);
+              int index;
+              switch (type) {
+                case TYOMAHDOLLISUUS -> index = tmLexicalIndex.getAndIncrement();
+                case KOULUTUSMAHDOLLISUUS -> index = kmLexicalIndex.getAndIncrement();
+                default -> throw new IllegalStateException("Unexpected value: " + type);
+              }
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java:157]: This seems to produce separate interleaved ordering for työ- and koulutusmahdollisuus.That is, given `t1,t2,...,k1,k2,...`, the index-based ordering will be `t1(index=0),k1(index=0),t2(1),k2(1),...` where (in lexical order) `t1 <= t2` and `k1 <= k2`, but t1 might be > k1.

Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java:157]: This comparison will been taken care at the frontend, when t and k arrays are combined and visible set of one page opportunities is sorted by the actual string.

Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java]: This has been changed to combined lexical index of all opportunities.

File: src/main/java/fi/okm/jod/yksilo/service/ehdotus/MahdollisuudetService.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Cacheable("mahdollisuusIdsAndTypes")
+  public LinkedHashMap<UUID, MahdollisuusTyyppi> fetchTyoAndKoulutusMahdollisuusIdsWithTypes(
+      Sort.Direction sort, Kieli lang) {
+    // fetch id's and title translation directly from translations
+    Query nativeQuery =
+        entityManager
+            .createNativeQuery(
+                SQL_UNION_OF_TYO_AND_KOULUTUSMAHDOLLISUUS_IDS + " " + sort, TypedResult.class)
```
[the remark this answers is not in the document]
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/service/ehdotus/MahdollisuudetService.java:60]: JPA does not support passing the "ORDER BY" clause directly as a parameter. This value is set from the enumeration.

File: src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java [hunk 4]
```diff
@@ -195,10 +220,11 @@ public record EhdotusMetadata(
       @Nullable Trendi trendi,
 
       /** Value from 0 to */
-      @Nullable Integer tyollisyysNakyma) {
+      @Nullable Integer tyollisyysNakyma,
+      @NotNull Integer lexicalOrder) {
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java:224]: Other attributes are in Finnish, so perhaps this should be `aakkosJarjestys` or something.

File: src/main/java/fi/okm/jod/yksilo/service/KoulutusmahdollisuusService.java [hunk 5]
```diff
@@ -38,6 +38,7 @@
 @RequiredArgsConstructor
 public class KoulutusmahdollisuusService {
   private final KoulutusmahdollisuusRepository koulutusmahdollisuudet;
+  private final EntityManager entityManager;
```
[the remark this answers is not in the document]
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/service/KoulutusmahdollisuusService.java]: Fixed

File: src/main/java/fi/okm/jod/yksilo/service/TyomahdollisuusService.java [hunk 6]
```diff
@@ -35,11 +35,7 @@
 @RequiredArgsConstructor
 public class TyomahdollisuusService {
   private final TyomahdollisuusRepository tyomahdollisuusRepository;
-
-  @Cacheable("tyomahdollisuusMetadata")
-  public Set<UUID> fetchAllIds() {
-    return tyomahdollisuusRepository.fetchAllIds();
-  }
+  private final EntityManager entityManager;
```
[the remark this answers is not in the document]
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/service/TyomahdollisuusService.java]: Fixed

File: src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java (hunk 4 above)
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/controller/ehdotus/MahdollisuudetController.java]: Changed

**Review by jhyoty — APPROVED** [review]

**Pull request merged**